[CDAP-18568] Support CustomVolumes and CustomVolumeMounts for CDAP System Services

dli357 opened the pull request:
This feature adds the ability for a user to directly specify custom k8s volumes and volume mounts in the CDAP CR.

Custom volumes and volume mounts specified in the upper-level CR will be applied to all services, and custom volumes and volume mounts specified in the particular service will only be applied to that service. Custom volume mounts for a service are applied to all containers and init containers for that service.

dvitiiuk commented:
Checked that this change satisfying the next requirements (which caused the creation of #72):
 - managing of file-based secure store on a separated volume
 - storing RBAC-related artifacts on a separated volume.

wyzhang commented on api/v1alpha1/cdapmaster_types.go:
```diff
@@ -93,6 +93,10 @@ type CDAPMasterSpec struct {
 	// SecurityContext defines the security context for all pods for all services.
 	SecurityContext *SecurityContext `json:"securityContext,omitempty"`
+	// CustomVolumes defines a list of custom volumes for the service.
+	CustomVolumes []corev1.Volume `json:"customVolumes,omitempty"`
```
Does this require yaml to specify an empty array like: 

CustomerVolumes: []

Or it is fine to omit CustomerVolumes in the yaml ?

wyzhang replied:
Nit: give some more context about this custom volume in the comment. A sample snippet of how it might be configured would be even better.

wyzhang replied:
nit: instead of calling these custom volumes, they are just additional volumes that are mounted on all services, if that is the case, just call them additionalVolumes. These volumes can PD, projected, or any types of volumes.  

"custom" makes it sound like a modified type of volume.

dli357 replied:
Switched to "AdditionalVolumes" and "AdditionalVolumeMounts" and added an additional comment pointing to the Kubernetes documentation on volumes.

Regarding empty lists, it is fine to omit the list entirely; the JSON or YAML deserialization should properly handle the null list case.

wyzhang commented on controllers/deployment.go:
```diff
@@ -326,6 +338,28 @@ func buildStatefulSetsObject(spec *StatefulSpec) (*reconciler.Object, error) {
+	if err := addVolumeToPodSpec(&statefulSetObj.Spec.Template.Spec, spec.Base.CustomVolumes); err != nil {
+		return nil, err
+	}
+	for ind, _ := range statefulSetObj.Spec.Template.Spec.InitContainers {
```
nit: ind => either idx or i. 

Have never seen "ind" as abbreviation for index

dli357 replied:
Fixed

wyzhang commented on controllers/deployment.go:
```diff
@@ -326,6 +338,28 @@ func buildStatefulSetsObject(spec *StatefulSpec) (*reconciler.Object, error) {
+			return nil, err
+		}
+	}
+	for ind, _ := range statefulSetObj.Spec.Template.Spec.Containers {
```
nit: same here ind=> idx

dli357 replied:
Fixed

wyzhang commented on controllers/deployment.go:
```diff
@@ -335,9 +369,55 @@ func buildDeploymentObject(spec *DeploymentSpec) (*reconciler.Object, error) {
+	if err := addVolumeToPodSpec(&deploymentObj.Spec.Template.Spec, spec.Base.CustomVolumes); err != nil {
+		return nil, err
+	}
+	for ind, _ := range deploymentObj.Spec.Template.Spec.InitContainers {
```
nit: same here and below ind=>idx

dli357 replied:
Fixed

wyzhang commented on controllers/testdata/logs.json:
```diff
@@ -172,16 +177,9 @@
                 "readOnly": true
               },
               {
-                "name": "cdap-cm-vol-my-config-map-1",
```
nit: is this a cleanup ? Cannot tell why it is related to your change

dli357 replied:
This is unrelated, but the init containers of the logging service should not have this volume mount. This likely slipped through (and still passes) because we are checking for the expected JSON template to be a superset of the output JSON, when in fact we should be checking the opposite for our integration tests. I've gone ahead and opened [CDAP-18732](https://cdap.atlassian.net/browse/CDAP-18732) to clean this up in the future.

wyzhang commented on go.mod:
```diff
@@ -3,12 +3,10 @@ module cdap.io/cdap-operator
 	github.com/nsf/jsondiff v0.0.0-20190712045011-8443391ee9b6
 	github.com/onsi/ginkgo v1.11.0
 	github.com/onsi/gomega v1.8.1
-	github.com/stretchr/testify v1.4.0
```
nit: are these changes needed?

dli357 replied:
These changes were included because I ran `go mod tidy`, so they should no longer be necessary anymore.

dli357 committed:
[CDAP-18568] Support CustomVolumes and CustomVolumeMounts for CDAP System Services

    [CDAP-18568] Switch name to AdditionalVolumes and AdditionalVolumeMounts and added clarifying comments
